tests/unit/lib/test_opensearch_exclusions.py

```diff
@@ -0,0 +1,221 @@
+        with self.harness.hooks_disabled():
+            self.charm.opensearch_exclusions.add("unit1")
+            self.charm.opensearch_exclusions.add("unit2")
+            self.charm.opensearch_exclusions.add("unit3")
+
+        assert sorted(self.charm.peers_data.get(Scope.APP, VOTING_TO_DELETE).split(",")) == [
+            "unit1",
+            "unit2",
+            "unit3",
+        ]
+
+    @patch(
+        "charms.opensearch.v0.opensearch_nodes_exclusions.OpenSearchExclusions._node",
+        new_callable=PropertyMock,
+    )
+    @patch("charms.opensearch.v0.opensearch_distro.OpenSearchDistribution.request")
+    @patch("charms.opensearch.v0.opensearch_nodes_exclusions.OpenSearchExclusions._fetch_voting")
+    def test_unit_restart_and_leadership_lost(self, mock_fetch, mock_request, mock_node_method):
+        """Test the change of leadership and how it impacts during a unit restart.
+
+        The flow works as follows:
+        1) Unit is the leader
+        2) Unit stops, in this test: add_current() is called
+        3) Unit loses leadership: in this case, the unit stops checking the Scope.APP,
+           switching to Scope.UNIT
+        4) Unit starts, in this test, delete_current() is called
+        """
+
+        class _node:  # noqa: N801
+            def is_cm_eligible(self):
+                return True
+
+            def is_data(self):
+                return True
+
+            @property
+            def name(self):
+                return "testing_unit"
+
+        mock_node_method.return_value = _node()
+        mock_request.return_value = 200
+
+        with self.harness.hooks_disabled():
+
+            # Unit is leader
+            self.harness.set_leader(True)
+            # Override the self._scope:
+            self.charm.opensearch_exclusions._scope = Scope.APP
+
+            # Populate the voting_to_delete with some units
+            self.charm.opensearch_exclusions.add("unit1")
+            self.charm.opensearch_exclusions.add("unit2")
+            self.charm.opensearch_exclusions.add("unit3")
+            # Simulating the unit stop
+            self.charm.opensearch_exclusions.add_current(voting=True, allocation=False)
+            mock_request.assert_called_with(
+                "POST",
+                "/_cluster/voting_config_exclusions?node_names=testing_unit&timeout=1m",
+                alt_hosts=[],
+                resp_status_code=True,
+                retries=3,
+            )
+            assert sorted(self.charm.peers_data.get(Scope.APP, VOTING_TO_DELETE).split(",")) == [
+                "testing_unit",
+                "unit1",
+                "unit2",
+                "unit3",
+            ]
+
+            # Unit loses leadership, the unit stops checking the Scope.APP
+            self.harness.set_leader(False)
+            # Override the self._scope:
+            self.charm.opensearch_exclusions._scope = Scope.UNIT
+            # Simulating the unit start, hence calling delete_current()
```

Comment:
I guess this is to be swapped with the line below, no? :-)

Reply:
Not really, the thing is, we never recreate the `self.charm.opensearch_exclusions` object, hence we never update its `_scope`. In theory, I do not even need to do `self.harness.set_leader(False)` to get the same result here.